tests/test_diff_rigid.py

```diff
@@ -0,0 +1,216 @@
+import numpy as np
+import pytest
+import torch
+
+import genesis as gs
+from genesis.utils.misc import ti_to_torch
+from genesis.engine.solvers.rigid.constraint_solver_decomp import func_init_solver, func_solve
+from genesis.engine.solvers.rigid.rigid_solver_decomp import kernel_step_1
+
+
+@pytest.mark.required
+@pytest.mark.field_only
+@pytest.mark.precision("64")
+@pytest.mark.parametrize("backend", [gs.cpu, gs.gpu])
+def test_diff_solver(backend):
+    torch.manual_seed(0)
+    rtol = 1e-4
+
+    scene = gs.Scene(
+        sim_options=gs.options.SimOptions(
+            dt=0.01,
+        ),
+        rigid_options=gs.options.RigidOptions(
+            # We use Newton's method because it converges faster than CG, and therefore gives better gradient estimation
+            # when using finite difference method
+            constraint_solver=gs.constraint_solver.Newton,
+        ),
+        show_viewer=False,
+    )
+
+    plane = scene.add_entity(gs.morphs.Plane(pos=(0, 0, 0)))
+    box = scene.add_entity(gs.morphs.Box(size=(1, 1, 1), pos=(10, 10, 0.49)))
```

Comment:
Prefer multiple lines in this context even if it fits in one line. It improves readability.